x-pack/platform/plugins/shared/alerting/server/lib/rule_gaps/task/gap_auto_fill_scheduler_task.ts

```diff
@@ -0,0 +1,417 @@
+
+                  if (!gapsPage.length) {
+                    break;
+                  }
+
+                  const filteredGaps = await filterGapsWithOverlappingBackfills(
+                    gapsPage,
+                    rulesClientContext,
+                    (message) => logger.warn(loggerMesage(message))
+                  );
+
+                  if (!filteredGaps.length) {
+                    if (gapsPage.length < gapsPerPage) {
+                      break;
+                    }
+                    continue;
+                  }
+
+                  const sortedGaps = filteredGaps.sort(
+                    (a, b) => a.range.gte.getTime() - b.range.gte.getTime()
+                  );
+                  const { results: chunkResults } = await processGapsBatch(rulesClientContext, {
+                    gapsBatch: sortedGaps,
+                    range: { start: startDate.toISOString(), end: now.toISOString() },
+                    initiator: backfillInitiator.SYSTEM,
+                    initiatorId: taskInstance.id,
+                  });
+                  addChunkResultsToAggregation(aggregatedByRule, chunkResults);
+
+                  // If fewer than gapsPerPage gaps returned, we've reached the end for this batch
+                  if (gapsPage.length < gapsPerPage) {
+                    break;
+                  }
+                }
+
+                // After finishing this rule batch, re-check backfill capacity
+                const capacityCheckPostBatch = await checkBackfillCapacity({
+                  rulesClient,
+                  maxBackfills: config.maxBackfills,
+                  logMessage: (message) => logger.warn(loggerMesage(message)),
+                  initiatorId: taskInstance.id,
+                });
+                if (!capacityCheckPostBatch.canSchedule) {
+                  const consolidated = resultsFromMap(aggregatedByRule);
+                  await logEvent({
+                    status: GAP_AUTO_FILL_STATUS.SUCCESS,
+                    results: consolidated,
+                    message: `Stopped early: no backfill capacity (${
+                      capacityCheckPostBatch.currentCount
+                    }/${capacityCheckPostBatch.maxBackfills}) | ${formatConsolidatedSummary(
+                      consolidated
+                    )}`,
+                  });
+                  return { state: {} };
+                }
+                // Update remaining capacity from the latest system state
+                remainingBackfills = capacityCheckPostBatch.remainingCapacity;
+              }
+
+              // Step 5: Finalize and log results
+              const consolidated = Array.from(aggregatedByRule.values());
+              if (consolidated.length === 0) {
+                await logEvent({
+                  status: GAP_AUTO_FILL_STATUS.SKIPPED,
+                  results: [],
+                  message: "Skipped execution: can't schedule gap fills for any enabled rule",
```

Comment:
can we add the reason why execution is skipped

Reply:
"can't schedule gap fills for any enabled rule" - shouldn't really happen, because first we get rule ids with unfilled gaps, then filter by enabled rules.

So, in theory we should always have something to schedule, but I added just as corner case